### PR TITLE
Cake 3.8 update

### PR DESCRIPTION
**Pull request by stefanorosanelli**

### CODE REVIEW

File: src/Application.php [hunk 1]
```diff
@@ -25,18 +27,69 @@
...
+    }
+
     /**
      * {@inheritDoc}
      */
     public function middleware($middlewareQueue) : MiddlewareQueue
     {
         $middlewareQueue
-            // Catch any exceptions in the lower layers,
+                    // Catch any exceptions in the lower layers,
```
Comment (batopa) [inline src/Application.php]: minor: remove extra spaces

File: src/Application.php [hunk 2]
```diff
@@ -45,8 +98,32 @@ public function middleware($middlewareQueue) : MiddlewareQueue
...
+    /**
+     * Get internal Csrf Middleware
+     *
+     * @return CsrfProtectionMiddleware
+     */
+    protected function csrfMiddleware() : CsrfProtectionMiddleware
+    {
+        // Csrf Middleware
+        $csrf = new CsrfProtectionMiddleware();
+            // Token check will be skipped when callback returns `true`.
```
Comment (batopa) [inline src/Application.php]: [minor] extra spaces

File: src/Application.php [hunk 3]
```diff
@@ -45,8 +98,32 @@ public function middleware($middlewareQueue) : MiddlewareQueue
...
+        $csrf->whitelistCallback(function ($request) {
+            $actions = (array)Configure::read(sprintf('CrsfExceptions.%s', $request->getParam('controller')));
```
Comment (batopa) [inline src/Application.php:114]: Maybe we could avoid CSRF skip using `X-CSRF-Token` in ajax request, see https://book.cakephp.org/3.0/en/controllers/middleware.html#csrf-protection-and-ajax-requests

I didn't understand if it is our case :upside_down_face:

**Review by batopa — CHANGES_REQUESTED** [review]

File: src/Application.php (hunk 1 above)
Comment (stefanorosanelli) [reply src/Application.php]: done

File: src/Application.php (hunk 2 above)
Comment (stefanorosanelli) [reply src/Application.php]: done

File: src/Template/Element/json_meta_config.ctp [hunk 4]
```diff
@@ -1,9 +1,9 @@
 <?php
     $csrfToken = null;
-    if (!empty($this->request->params['_csrfToken'])) {
-        $csrfToken = json_encode($this->request->params['_csrfToken']);
+    if (!empty($this->request->getParam('_csrfToken'))) {
+        $csrfToken = json_encode($this->request->getParam('_csrfToken'));
     } elseif (!empty($this->request->data['_csrfToken'])) {
-        $csrfToken = json_encode($this->request->data['_csrfToken']);
+        $csrfToken = json_encode($this->request->getParam('_csrfToken'));
```
Comment (didoda) [inline src/Template/Element/json_meta_config.ctp]: typo?
it should be `$csrfToken = json_encode($this->request->getData('_csrfToken'));` or somethink like that...

**Review by batopa — APPROVED** [review]

**Review by didoda — APPROVED** [review]

**Pull request merged**